sonar-python-plugin/src/main/java/org/sonar/plugins/python/xunit/TestSuiteParser.java

```diff
@@ -19,39 +19,53 @@
  */
 package org.sonar.plugins.python.xunit;
 
-import org.codehaus.staxmate.in.ElementFilter;
-import org.codehaus.staxmate.in.SMHierarchicCursor;
-import org.codehaus.staxmate.in.SMInputCursor;
-import org.sonar.api.utils.ParsingUtils;
-import org.sonar.api.utils.StaxParser.XmlStreamHandler;
-
-import javax.xml.stream.XMLStreamException;
+import java.io.File;
 import java.text.ParseException;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Optional;
+import javax.xml.stream.XMLStreamException;
+import javax.xml.xpath.XPathConstants;
+import org.sonar.api.utils.ParsingUtils;
+import org.sonar.plugins.python.PythonReportException;
+import org.sonar.plugins.python.XmlReportParser;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
 
-public class TestSuiteParser implements XmlStreamHandler {
+public class TestSuiteParser extends XmlReportParser {
 
   private Map<String, TestSuite> testSuites = new HashMap<>();
 
-  @Override
-  public void stream(SMHierarchicCursor rootCursor) throws XMLStreamException {
-    SMInputCursor testSuiteCursor = rootCursor.constructDescendantCursor(new ElementFilter("testsuite"));
-    while (testSuiteCursor.getNext() != null) {
-      String testSuiteClassName = testSuiteCursor.getAttrValue("name");
-
-      SMInputCursor testCaseCursor = testSuiteCursor.childElementCursor("testcase");
-      while (testCaseCursor.getNext() != null) {
-        String testClassName = getClassname(testCaseCursor, testSuiteClassName);
-        TestSuite report = testSuites.get(testClassName);
-        if (report == null) {
-          report = new TestSuite(testClassName);
-          testSuites.put(testClassName, report);
+  public void parse(File xmlFile) {
+    try {
+      Document xmlDocument = getDocumentBuilder().parse(xmlFile);
+
+      NodeList testsuiteNodes = (NodeList) getXpath().compile("testsuite").evaluate(xmlDocument, XPathConstants.NODESET);
+
+      for (int i = 0; i < testsuiteNodes.getLength(); i++) {
+        Element testsuiteNode = (Element) testsuiteNodes.item(i);
+
+        Optional<String> testSuiteClassName = getStringAttribute(testsuiteNode, "name");
+
+        NodeList testcaseNodes = testsuiteNode.getElementsByTagName("testcase");
+        for (int j = 0; j < testcaseNodes.getLength(); j++) {
+          Element testcaseNode = (Element) testcaseNodes.item(j);
+          String testClassName = getClassname(testcaseNode, testSuiteClassName.orElse(null));
+
+          TestSuite report = testSuites.get(testClassName);
+          if (report == null) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)